### ISSUE TITLE
Advanced Effects - Theory vs. Practice

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#1080]
The Flix effect system allows us to track whether a function is pure or impure.

For example:

```scala
def isEmpty(o: Option[a]): Bool & Pure = ... // pure is the default and can be omitted
```
and:
```scala
def print(s: String): Unit & Impure = ...
```

We can *demand* a pure or impure function. For example:

```scala
def count(f: a -> Bool, xs: List[a]): Int32 = ... // requires a pure function
def unfoldWithIter(next: () ~> Option[a]): List[a] & Impure = ... // requires an impure function
```

We can also express that functions are *effect polymorphic* - i.e. their purity depends on the purity of their argument(s). For example:

```scala
def map(f: a -> b & e, o: Option[a]): Option[b] & e = ...
```

We can even express when a function is effect polymorphic in multiple higher-order functions:

```scala
def >>(f: a -> b & e1, g: b -> c & e2): a -> c & {e1, e2} = ...
```

That is, function composition is *pure* if both `e1` and `e2` are pure. The `{e1, e2}` fragment can be read as `e1 /\ e2` (i.e. logical and).

I think that the last couple of weeks have shown that this design works reasonably well. Separating pure and impure code should make it easier to reason about programs and also to avoid stupid mistakes (e.g. passing a pure function to `unfoldWithIter`). In the future, we can leverage purity and impurity information to build more expressive features and potentially for compiler optimizations. 

**CrazyModeOn.**

The effect system in Flix is unique. It is developed as part of a research project that combines Hindley-Milner style type inference with boolean unification. We are currently writing a paper about the scientific results and hope to submit it soon. 

The effect system is significantly more expressive that the above examples show. In fact, we can use *arbitrary boolean formulas* to constrain purity. 

For example, we can write:

```scala
def requiresPureAndImpure1(f: a -> b & e, g: a -> b & {{not e}}): Unit = ...
```

This function requires *exactly one pure and one impure function argument*. If `f` is pure then `g` *must be* impure and vice versa! 

As another example, we can write:

```scala
def requiresAtmostOnePure1(f: a -> b & e1, g: a -> b & {{not e1 /\ e2}}): Unit = ...
```

This function requires *at most one pure function*. That is, you can call it with a pure and an impure function, an impure and a pure function, or two impure functions. 

This system is a bit "crazy". For example, we can write the program:

```scala
def pure(): Unit = ()
def yikes(f: a -> b & {{(x \/ y \/ z) /\ (x \/ y \/ not z) /\ (x \/ not y \/ z) /\ (x \/ not y \/ not z) /\ (not x \/ y \/ z) /\ (not x \/ y \/ not z) /\ (not x \/ not y \/ z) /\ (not x \/ not y \/ not z)}}): Unit = ()
def testYikes(): Unit = yikes(pure)
```

This program type checks if the 3SAT CNF formula:

```
(x \/ y \/ z) /\ (x \/ y \/ not z) /\ (x \/ not y \/ z) /\ (x \/ not y \/ not z) /\ (not x \/ y \/ z) /\ (not x \/ y \/ not z) /\ (not x \/ not y \/ z) /\ (not x \/ not y \/ not z)
```
has a solution.

Now, another "strange" thing is that the two functions:

```scala
def requiresPureAndImpure1(_f: a -> b & e, _g: a -> b & {{not e}}): Unit = ...
def requiresPureAndImpure2(_f: a -> b & {{not e}}, _g: a -> b & e): Unit = ...
```
actually have the same type signature!

That is, types are equal modulo boolean equivalences.

**CrazyModeOff.**

If the latter seems crazy, it is important to understand that it is the formal foundation for the simpler use cases in the start of the issue. Now, the important point is: *We do not have to expose such crazy types as shown in the latter part*. We get to decide what should actually be available to the programmer.

You might wonder: Do you have any legitimate use cases for the crazy stuff? We have one good use case. We want to write a lint that expresses that two map operations can be merged into one map operation if *at most* one is impure. Concretely we could imagine something like:

```scala
@lint
def mapMap1(): Bool =
    forall [a, b, c] (f: a -> b, g: b -> c, xs: List[a]).
        map(g, map(f, xs)) === map(f >> g, xs)

@lint
def mapMap2(): Bool =
    forall [a, b, c] (f: a ~> b, g: b -> c, xs: List[a]).
        map(g, map(f, xs)) === map(f >> g, xs)

@lint
def mapMap3(): Bool =
    forall [a, b, c] (f: a -> b, g: b ~> c, xs: List[a]).
        map(g, map(f, xs)) === map(f >> g, xs)
```

This handles the case where both `f` and `g` are pure, where `f` is impure, and `g` is impure. With the crazy scheme above, we can write just one lint:

```scala
@lint
def mapMap(): Bool =
    forall [a, b, c] (f: a -> b & e1, g: b -> c & {{not e1 \/ e2}}, xs: List[a]).
        map(g, map(f, xs)) === map(f >> g, xs)
```

Design Space:

- Expose the entire system to the programmer, but keep the awkward `{{...}}}` syntax to highlight that something exceptional is going on.
- Expose the entire system to the programmer and improve the syntax.
- Expose only the parts of the system required for the use cases in the beginning of the issue (i.e. disallow boolean formulas, only allow variables and conjunction).

**TLDR:** Flix has a crazy effect system based on arbitrary boolean formulas. How much of this should be exposed to the user?

**Comment by mlutze** [comment]
We could consider an actual `crazyModeOn` compiler flag, comparable to Scala's advanced feature flags.

**Comment by timjs** [comment]
Interesting thought! Would these kind of annotations be used by the compiler to optimize code by rewriting? Akin [Haskell's rewrite rules](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#rewrite-rules)?

**Comment by magnus-madsen** [comment]
> 
> 
> Interesting thought! Would these kind of annotations be used by the compiler to optimize code by rewriting? Akin [Haskell's rewrite rules](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#rewrite-rules)?

Yes. 

We plan to have several "lints" where one lint can be an optimization (i.e. rewrite rule applied by the compiler) and another can be "this code looks wrong" and a third could be "this code can be simplified to".
